platform-sdk/swirlds-common/src/main/java/com/swirlds/common/merkle/hash/MerkleHashBuilder.java

```diff
@@ -142,77 +113,104 @@ public Future<Hash> digestTreeAsync(MerkleNode root) {
 
     /**
      * The root of a merkle tree.
      *
      * @param it
      * 		An iterator that walks through the tree.
-     * @param activeThreadCount
-     * 		If single threaded then this is null, otherwise contains the number of threads
-     * 		that are actively hashing. Once the active thread count dips below the maximum,
-     * 		this means that one thread has either finished or exploded.
      */
-    private void hashSubtree(final Iterator<MerkleNode> it, final AtomicInteger activeThreadCount) {
+    private void hashSubtree(final Iterator<MerkleNode> it) {
         while (it.hasNext()) {
             final MerkleNode node = it.next();
-            // Potential optimization: if this node is currently locked, do not wait here. Skip it and continue.
-            // This would require a lock object that support the "try lock" paradigm.
-            synchronized (node) {
-                if (activeThreadCount != null && activeThreadCount.get() != cpuThreadCount) {
-                    break;
-                }
 
-                if (node.getHash() != null) {
-                    continue;
-                }
+            if (node.getHash() != null) {
+                continue;
+            }
+
+            merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+        }
+    }
+
+    class TraverseTask extends AbstractTask {
+        final MerkleNode node;
+        final AbstractTask out;
 
-                if (node.isLeaf()) {
-                    merkleCryptography.digestSync(node.asLeaf(), MERKLE_DIGEST_TYPE);
+        TraverseTask(final MerkleNode node, AbstractTask out) {
+            super(threadPool, 1);
+            this.node = node;
+            this.out = out;
+        }
+
+        @Override
+        protected boolean exec() {
+            try {
+                if (node == null || node.getHash() != null) {
+                    out.send();
+                } else if (node.isLeaf()) {
+                    merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+                    out.send();
                 } else {
-                    merkleCryptography.digestSync(node.asInternal(), MERKLE_DIGEST_TYPE);
+                    MerkleInternal internal = node.asInternal();
+                    int nChildren = internal.getNumberOfChildren();
+                    if (nChildren == 0) {
+                        merkleCryptography.digestSync(node, MERKLE_DIGEST_TYPE);
+                        out.send();
+                    } else {
+                        ComputeTask compute = new ComputeTask(internal, nChildren, out);
+                        for (int childIndex = 0; childIndex < nChildren; childIndex++) {
+                            MerkleNode child = internal.getChild(childIndex);
+                            new TraverseTask(child, compute).send();
+                        }
+                    }
                 }
+            } catch (Exception ex) {
+                resultFuture.cancelWithException(ex);
```

Comment:
I wonder if `out.completeExceptionally()` can be used here? It would propagate the exception up to the ResultTask, which will then call `cancelWithError()` on the result future. It would eliminate need to have `resultFuture` as a field, it can be just a local var

Reply:
Done